api/envoy/extensions/transport_sockets/starttls/v3/starttls.proto

```diff
@@ -21,15 +21,28 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // a network filter which understands protocol exchange and a state machine to signal to the StartTls
 // transport socket when a switch to TLS is required.
 
-// Configuration for StartTls transport socket.
+// Configuration for a downstream StartTls transport socket.
 // StartTls transport socket wraps two sockets:
-// - raw_buffer socket which is used at the beginning of the session
-// - TLS socket used when a protocol negotiates a switch to encrypted traffic.
+// * raw_buffer socket which is used at the beginning of the session
+// * TLS socket used when a protocol negotiates a switch to encrypted traffic.
 message StartTlsConfig {
   // (optional) Configuration for clear-text socket used at the beginning of the session.
   raw_buffer.v3.RawBuffer cleartext_socket_config = 1;
 
-  // Configuration for TLS socket.
+  // Configuration for a downstream TLS socket.
   transport_sockets.tls.v3.DownstreamTlsContext tls_socket_config = 2
       [(validate.rules).message = {required: true}];
 }
+
+// Configuration for an upstream StartTls transport socket.
+// StartTls transport socket wraps two sockets:
+// * raw_buffer socket which is used at the beginning of the session
+// * TLS socket used when a protocol negotiates a switch to encrypted traffic.
```

Comment:
How this config is supposed to be used? For downstream starttls config, it have to be paired with a network filter supporing it (e.g. postgresql), for upstream, which filter supports that? Can you add more documentation here?

Reply:
The documentation on lines 19-22 are what you're asking for. I assumed they would be understood as shared between the two different configurations, but I'm happy to duplicate it if you'd like.

Reply:
My question here is do we have or have a plan of a network filter supporting upstream starttls, without that this shouldn't be configured at anytime, and this part of document will just confuses user.

Reply:
Who is the 'we' in this scenario? Twitter has an interest in it to support mux opportunistic-tls. Whether this protocol implementation will make it upstream into envoy is undecided, but probably leaning toward 'no' at this time since ideally we'd get out of the home-grown protocol business.

Reply:
'We' means the project here. I understand that you don't want to upstream the protocol implementation that Twitter uses internally and that's totally ok. Just trying to lay out what is the plan going forward, and not confuse Envoy users with the documentation.

Reply:
I took a swing at elaborating. Let me know what you think.

Reply:
lgtm